### ISSUE TITLE
False positive when applying a decorator using an union with a bounded typevar

### ISSUE THREAD

**Issue by Wizzerinus** [astral-sh/ty#3022]
### Summary

The following code emits an error on the `@deco`:
```py
from collections.abc import Callable
from typing import ParamSpec, TypeVar

T = TypeVar("T")
P = ParamSpec("P")
C = TypeVar("C", bound=type)

def deco(f: Callable[P, T] | C) -> Callable[P, T] | C:
    return f


@deco
# Error: Argument to function `deco` is incorrect:
# Argument type `def someFunc() -> Unknown` does not satisfy upper bound `type` of type variable `C`
def someFunc():
    pass
```
However, it seems to me that this is clearly reasonable, since removing the union from the parameter type and making it just `Callable[P, T] -> Callable[P, T] | C` or similar makes it not emit any errors. Replacing the `P` with an ellipsis also makes the error go away.

Playground link: https://play.ty.dev/2fe6911c-e65b-4f20-ba72-032829128517

### Version

ty 0.0.21

**Comment by sharkdp** [comment]
Thank you for reporting this.

Our support for unions in the generics solver is limited at the moment. See https://github.com/astral-sh/ty/issues/623 and https://github.com/astral-sh/ty/issues/1703 for related tickets. I'm marking this as a child of #623 to make sure it gets eventually included as a test case.

**Comment by andyli-oai** [comment]
Another minimal example that appears related, reproduced on **ty 0.0.73**:

```python
from typing import Any, TypeVar

TNumber = TypeVar("TNumber", bound=int)


def identity(value: TNumber | dict[str, Any]) -> TNumber | dict[str, Any]:
    return value


def caller(value: int | dict[str, Any]) -> None:
    identity(value)
```

Running `ty check repro.py` reports `invalid-argument-type` on `identity(value)`:

```text
Argument to function `identity` is incorrect:
Argument type `int | dict[str, Any]` does not satisfy
upper bound `int` of type variable `TNumber`
```

Expected: no error. `TNumber` can specialize to `int`, while the dictionary matches the separate union alternative.

**Comment by carljm** [comment]
Yes, these cases are both more precisely related to #3557, which I'm actively working on.
